source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,462 @@
+[transition]: https://github.com/alphagov/transition
+[transition-config]: https://github.com/alphagov/transition-config
+
+### Routing on GOV.UK
+
+#### Getting to the 'router' application
+
+![](images/load-balancer.png)
+
+Some requests make it through the CDN and cache layers to 'origin'. Origin is
+a stack of computers in the cloud - in this case, AWS - and its entry point is
+a load balancer.
+
+The load balancer knows based on the hostname which machine 'class' to route
+to. Different classes of machine run different sets of GOV.UK applications.
+How many machines are allocated to a class - and how big those machines are -
+is configured using [Terraform], in [govuk-aws]. What runs on each machine
+class is configured in govuk-puppet, a file and process management system we'll
+cover in more detail later.
+
+![](images/nginx-routing.png)
+
+External requests are routed to a 'cache' machine, which uses Varnish for its
+caching. If Varnish has the route response in its cache, it returns that,
+otherwise it forwards the request to [Nginx], a web server running on the
+machine. Nginx is configured with govuk-puppet.
+
+Nginx proxies most requests to [router], which is a GOV.UK maintained
+application running on the cache machines. However, Nginx does proxy some
+routes directly to other apps, such as image URLs routing to asset-manager.
+
+[govuk-aws]: https://github.com/alphagov/govuk-aws
+[Nginx]: https://www.nginx.com/
+[router]: https://github.com/alphagov/router
+[Terraform]: https://www.terraform.io/
+
+#### Routing via 'router'
+
+'router' is a reverse proxy app written in Go. It stores all known routes
+in memory using a [prefix trie]. The routes are loaded into the trie from
+[redis] at startup. Matched routes are proxied through to the relevant
+rendering app based on the `backend_id` of the route, which is derived from
+the `rendering_app` field in the corresponding content item in the content
+store (which we'll cover later). For example, if the route has a `backend_id`
+of `frontend`, it will forward the request to the [frontend] application.
+
+When routes are deleted, they're marked as `gone` in router, and router
+returns a 410 Gone response. When routes are changed (which can happen when a
+document's title is changed, affecting its slug), it is marked as `redirect`
+in router, serving a 301 Moved Permanently response. It's worth noting the
+[short-url-manager] app, whose sole purpose is to allow the creation of short,
+memorable URLs that redirect to longer URLs, often as part of a media campaign.
+It publishes routes to the content store (via the publishing API, which we'll
+also cover later) like other apps publish pages.
+
+Routes are added and updated in real time via [router-api]. The content
+store talks to router-api [directly][register-route] whenever there is a new
+piece of content router needs to know about. router-api stores this information
+in its own database and then sends these routes to router.
```

Comment:
It would be more correct to say "The Router API stores this information in it's database, and then sends requests to the instances of the Router to reload the routes from the database".

One interesting property here is that because there are multiple router processes, one for each cache machine, the Router API has to contact each of them to tell them to reload the routes. This means the Router API has to know about where to find each individual cache machine instance

Reply:
Thanks Chris. Instead of that, should it be this?

> The Router API stores this information in it's database, and then sends requests to the Router on every 'cache' machine to reload the route.

What you say about router reloading routes from the database contradicts what I've heard from others. My understanding is that it is _indirectly_ loaded from the database at startup, i.e. router-api loads routes from MongoDB and pushes the routes to Router; Router doesn't know about the MongoDB. And then if routes are added/changed at runtime, they're individually added/overwritten via router-api. I may have this wrong though.

Reply:
> Thanks Chris. Instead of that, should it be this?
> 
> > The Router API stores this information in it's database, and then sends requests to the Router on every 'cache' machine to reload the route.

As far as I know, the Router can only be asked to reload all the routing information, not a specific route.

> What you say about router reloading routes from the database contradicts what I've heard from others. My understanding is that it is _indirectly_ loaded from the database at startup, i.e. router-api loads routes from MongoDB and pushes the routes to Router; Router doesn't know about the MongoDB. And then if routes are added/changed at runtime, they're individually added/overwritten via router-api. I may have this wrong though.

As I understand it, the Router and Router API both access a MongoDB database. The Router API reads and writes, and I think the Router only reads. When the Router API changes something in the database, it sends an empty POST request [1] to each of the Router processes, triggering to reload all the routing information from that shared MongoDB database.

1: https://github.com/alphagov/router-api/blob/master/lib/router_reloader.rb#L45-L48

Reply:
Thanks Chris. I've updated the docs accordingly 👍